pkg/virt-launcher/virtwrap/network/podinterface.go

```diff
@@ -69,8 +69,84 @@ type BindMechanism interface {
 	startDHCP(vmi *v1.VirtualMachineInstance) error
 }
 
-type podNICImpl struct {
-	handler NetworkHandler
+// Network configuration is split into two parts, or phases, each executed in a
+// different context.
+// Phase1 is run by virt-handler and heavylifts most configuration steps. It
+// also creates the tap device that will be passed by name to virt-launcher,
+// thus allowing unprivileged libvirt to consume a pre-configured device.
+// Phase2 is run by virt-launcher in the pod context and completes steps left
+// out of virt-handler. The reason to have a separate phase for virt-launcher
+// and not just have all the work done by virt-handler is because there is no
+// ready solution for DHCP server startup in virt-handler context yet. This is
+// a temporary limitation and the split is expected to go once the final gap is
+// closed.
+// Moving all configuration steps into virt-handler will also allow to
+// downgrade privileges for virt-launcher, specifically, to remove NET_ADMIN
+// capability. Future patches should address that. See:
+// https://github.com/kubevirt/kubevirt/issues/3085
+type podNIC struct {
+	vmi              *v1.VirtualMachineInstance
+	podInterfaceName string
+	iface            *v1.Interface
+	network          *v1.Network
+	handler          NetworkHandler
+}
+
+func newPodNIC(vmi *v1.VirtualMachineInstance, network *v1.Network, handler NetworkHandler) (*podNIC, error) {
+	if network.Pod == nil && network.Multus == nil {
+		return nil, fmt.Errorf("Network not implemented")
+	}
+
+	correspondingNetworkIface := findInterfaceByNetworkName(vmi, network)
+	if correspondingNetworkIface == nil {
+		return nil, fmt.Errorf("no iface matching with network %s", network.Name)
+	}
+
+	composedPodInterfaceName, err := composePodInterfaceName(vmi, network)
+	if err != nil {
+		return nil, err
+	}
+	return &podNIC{
+		handler:          handler,
+		vmi:              vmi,
+		network:          network,
+		podInterfaceName: composedPodInterfaceName,
+		iface:            correspondingNetworkIface,
+	}, nil
+}
+
+func composePodInterfaceName(vmi *v1.VirtualMachineInstance, network *v1.Network) (string, error) {
+	if network.Multus != nil && !network.Multus.Default {
+		multusIndex := findMultusIndex(vmi, network)
+		if multusIndex == nil {
+			return "", fmt.Errorf("Network name %s not found", network.Name)
+		}
+		return fmt.Sprintf("net%d", *multusIndex), nil
+	}
+	return primaryPodInterfaceName, nil
+}
+
+func findInterfaceByNetworkName(vmi *v1.VirtualMachineInstance, network *v1.Network) *v1.Interface {
+	for i, iface := range vmi.Spec.Domain.Devices.Interfaces {
+		if iface.Name == network.Name {
+			return &vmi.Spec.Domain.Devices.Interfaces[i]
+		}
+	}
+	return nil
+}
+
+func findMultusIndex(vmi *v1.VirtualMachineInstance, networkToFind *v1.Network) *int {
```

Comment:
I personally find it better if you return an int , with -1 to represent not found. Since https://golang.org/pkg/strings/#Index is designed like that, I find it more idiomatic. 

Will not push for it, it's just an opinion.

Reply:
Let's do it idiomatic.